### ISSUE TITLE
[slim-engine-dev] more internal checks

### ISSUE THREAD

**Issue by grahamgower** [popsim-consortium/stdpopsim#265]
* Fail earlier for impossible sampling. I.e. cannot ask for n samples, where n > N/Q.
* ~~Verify that the final tree sequence object has the requested number of sampled individuals for each population. Or maybe this is something that should be checked after Engine.simulate() is called?~~

Probably other things.

**Comment by jeromekelleher** [comment]
> Verify that the final tree sequence object has the requested number of sampled individuals for each population. Or maybe this is something that should be checked after Engine.simulate() is called?

This sounds like a unit test to me rather than a standard run-time check.

**Comment by grahamgower** [comment]
Yeah, I think it already is.

**Comment by grahamgower** [comment]
Requests for sample sizes greater than N/Q currently do not give an error (but you only get N/Q 
individuals back).

```
$ stdpopsim -e slim --slim-scaling-factor 100 HomSap -l 0.001 1000 -o n1k.trees -q
$ tskit info n1k.trees
sequence_length:  249250.0
trees:            645
samples:          200
individuals:      100
nodes:            782
edges:            2550
sites:            823
mutations:        823
migrations:       0
populations:      1
provenances:      5
```

**Comment by grahamgower** [comment]
Resolved with #478.
